ydb/services/metadata/manager/generic_manager.h

```diff
@@ -42,35 +44,123 @@ class TGenericOperationsManager: public IObjectOperationsManager<T> {
+        }
+        if (!manager->GetOperationsManager()) {
+            return TYqlConclusionStatus::Fail("modification is unavailable for " + manager->GetTypeId());
+        }
+
+        TOperationParsingResult patch(TBase::BuildPatchFromSettings(settings, context));
+        if (!patch.IsSuccess()) {
+            return TYqlConclusionStatus::Fail(patch.GetErrorMessage());
+        }
+
+        NKqpProto::TKqpPhyMetadataOperation* operationProto = nullptr;
+        switch (context.GetActivityType()) {
+            case EActivityType::Upsert:
+                operationProto = schemeOperation.MutableUpsertObject();
+                break;
+            case EActivityType::Create:
+                operationProto = schemeOperation.MutableCreateObject();
+                operationProto->SetSuccessOnAlreadyExists(settings.GetExistingOk());
+                break;
+            case EActivityType::Alter:
+                operationProto = schemeOperation.MutableAlterObject();
+                break;
+            case EActivityType::Drop:
+                operationProto = schemeOperation.MutableDropObject();
+                operationProto->SetSuccessOnNotExist(settings.GetMissingOk());
+                break;
+            case EActivityType::Undefined:
+                return TYqlConclusionStatus::Fail("undefined action type");
+        }
+        Y_DEBUG_ABORT_UNLESS(operationProto);
+
+        for (const auto& [col, value] : patch.GetResult().GetValues()) {
+            auto* proto = operationProto->AddColumnValues();
+            proto->SetColumn(col);
+            proto->MutableValue()->CopyFrom(value);
+        }
+
+        return TYqlConclusionStatus::Success();
+    }
+
+    virtual NThreading::TFuture<TYqlConclusionStatus> ExecutePrepared(const NKqpProto::TKqpSchemeOperation& schemeOperation,
+        const ui32 nodeId, const IClassBehaviour::TPtr& manager, const TExternalModificationContext& context) const override {
+        if (!manager) {
+            return NThreading::MakeFuture<TYqlConclusionStatus>(TYqlConclusionStatus::Fail("modification object behaviour not initialized"));
+        }
+        if (!manager->GetOperationsManager()) {
+            return NThreading::MakeFuture<TYqlConclusionStatus>(TYqlConclusionStatus::Fail("modification is unavailable for " + manager->GetTypeId()));
+        }
+
+        auto promise = NThreading::NewPromise<TYqlConclusionStatus>();
+        auto controller = std::make_shared<TOperationsController>(promise);
+        IObjectModificationCommand::TPtr modifyObjectCommand;
+        TInternalModificationContext internalContext(context);
+        switch (schemeOperation.GetOperationCase()) {
+            case NKqpProto::TKqpSchemeOperation::kUpsertObject:
+                internalContext.SetActivityType(EActivityType::Upsert);
+                modifyObjectCommand = std::make_shared<TUpsertObjectCommand<T>>(BuildPatchFromProto(schemeOperation.GetUpsertObject()), manager, controller, internalContext);
```

Comment:
We make a fresh TOperationsController pointer for each ObjectCommand, and we copy it inside. Also we copy it in TInsertObjectsActor constructor. Do we really need those copies, or maybe we could use unique pointer? Can ObjectCommand remain, if corresponding TOperationsController is destroyed?

Reply:
I only refactored this code and I don't know why it is shared. But as I see, it can be simplified as you suggested. Thank you, I will try :)

Reply:
I tried to use unique_ptr and noticed it: https://github.com/ydb-platform/ydb/blob/main/ydb/services/metadata/initializer/accessor_init.h#L27
I am afraid that we need shared_ptr :(

Reply:
I avoided extra copying here using std::move - it is all that I can do :)